### PR TITLE
Added: Generator for the full implementation

### PR DESCRIPTION
**Pull request by xescugc**
As all the implementations are the same, I've created a generator that from a configuration it's able to create the needed Interface and implementation

### CODE REVIEW

**Comment by xescugc** [comment]
The tests did decrease because all the generated code does not have tests, as It would mean to generate tests too, basically because the tests where also the same for everyone of the methods.

File: cmd/template.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+		panic(err)
+	}
+
+	awsReaderTmpl, err = template.New("test").Parse(arTmpl)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Function it's the definition of one of the functions
```
Comment (xlr-8) [inline cmd/template.go]: is the definition

File: cmd/template.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	// FnSignature it's the signture it has to be used on the Interface
+	// AWSReader and the function implementation
+	FnSignature string
+
+	// NoGenerateFn avoids generating the function implementation as
+	// it's to off of the pattern
+	// If true, it should be used with 'Signature' to add it to the
+	// AWSReader and have the custom implementation outside of the
+	// generated code
+	NoGenerateFn bool
```
Comment (xlr-8) [inline cmd/template.go:148]: is the name* x2
is used*
is the AWS service*
is the documentation*
is the output name*
is the signature used by the Interface*

avoids generating the function as it's no respecting the signature pattern?

File: cmd/template.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+// Name builds a name simply using "Get{{.Entity}}"
+// except if FnName it's defined, in which case
+// only FnName it's used
```
Comment (xlr-8) [inline cmd/template.go]: is defined, is used

File: cmd/template.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	if f.FnName != "" {
+		return f.FnName
+	}
+
+	return fmt.Sprintf("Get%s", f.Entity)
+}
+
+// Output builds the output by "{{.Service}}.{{.Prefix}}{{.Entity}}"
+// except if FnOutput it's defined in which case the formula
+// "{{.Service}}{{.FnOutput}}" it's used
```
Comment (xlr-8) [inline cmd/template.go]: is defined, is used

File: cmd/template.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+	return fmt.Sprintf("%s.%s%sOutput", f.Service, f.Prefix, f.Entity)
+}
+
+// Input builds the input by "{{.Service}}.{{.Prefix}}{{.Entity}}"
+func (f Function) Input() string {
+	return fmt.Sprintf("%s.%s%sInput", f.Service, f.Prefix, f.Entity)
+}
+
+// Signature builds the signature except if FnSignature it's defined,
```
Comment (xlr-8) [inline cmd/template.go:187]: is defined

File: cmd/functions.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
+package main
+
+var (
+	// functions it's the list of fuctions that will be added
```
Comment (xlr-8) [inline cmd/functions.go]: is the list

File: cmd/functions.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+			// GetVolumes returns all EC2 volumes based on the input given.
+			// Returned values are commented in the interface doc comment block.
+			`,
+		},
+		Function{
+			Entity:  "Snapshots",
+			Prefix:  "Describe",
+			Service: "ec2",
+			Documentation: `
+			// GetSnapshots returns all snapshots belonging to the Account ID based on the input given.
```
Comment (xlr-8) [inline cmd/functions.go]: If this code is generated, then I don't think it returns what belonged to you anymore.

That's a behaviour made sense to only have snapshots/AMI belonging to the caller (especially if it's about seeing what's yours/importing, etc), but in a more general/public point of view, we might want to move that with a separated function like some of the S3 ones.

File: cmd/functions.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+			// Returned values are commented in the interface doc comment block.
+			`,
+		},
+
+		// elasticache
```
Comment (xlr-8) [inline cmd/functions.go:92]: are those commentary still relevant?

**Review by xlr-8 — CHANGES_REQUESTED** [review]

File: cmd/template.go (hunk 2 above)
Comment (xescugc) [reply cmd/template.go:148]: > avoids generating the function as it's no respecting the signature pattern?

For now yes, but could be different, maybe some list requires something to different to what we are templating now.

File: cmd/functions.go (hunk 8 above)
Comment (xescugc) [reply cmd/functions.go:92]: It's to keep the list somehow grouped, with AWS Services.

File: cmd/functions.go (hunk 7 above)
Comment (xlr-8) [reply cmd/functions.go]: As noted by @xescugc, same goes for AMI. It was decided to implement both for flexibility reasons.

**Comment by xescugc** [comment]
Ok pushed the changes.

**Review by xlr-8 — APPROVED** [review]

File: cmd/functions.go (hunk 8 above)
Comment (xlr-8) [reply cmd/functions.go:92]: OK

**Comment by xescugc** [comment]
Added test and some docs on the Readme

File: README.md [hunk 9]
```diff
@@ -50,6 +50,12 @@ fmt.Println(vpcs)
 
 You can also take a look at the [example file](example/main.go).
 
+### Contribute
+
+We use a custom generation tool located on `cmd/main.go` which basically uses a list of function definitions (`cmd/functions.go`) to generate the wrappers for those, if you want to add a call to the AWS API you have to add it to that list and if the implementation fits the template it'll be automatically generated/implemented. If it does not fit the template you'll have to implement it manually.
```
Comment (xlr-8) [inline README.md]: Maybe make it multi-line? And show an example of the manual implementation (AMI/S3)?

**Comment by xlr-8** [comment]
Should we have some tests regarding already existing method? Like if 2 signatures collide because one is implemented manually the other generated?

**Comment by xescugc** [comment]
BTW the tests are using `testify` https://github.com/cycloidio/raws/issues/32

**Review by xlr-8 — APPROVED** [review]

**Comment by xlr-8** [comment]
LGTM! :+1:

**Pull request merged**